src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -1107,4 +1113,181 @@ public LeafRope flatten(Rope rope) {
 
     }
 
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "a"),
+            @NodeChild(type = RubyNode.class, value = "b")
+    })
+    public abstract static class EqualNode extends RubyNode {
+
+        public static EqualNode create() {
+            return RopeNodesFactory.EqualNodeGen.create(null, null);
+        }
+
+        public abstract boolean execute(Rope a, Rope b);
+
+        @Specialization(guards = "a == b")
+        public boolean sameRopeEqual(Rope a, Rope b) {
+            return true;
+        }
+
+        @Specialization
+        public boolean ropesEqual(Rope a, Rope b,
+                @Cached("create()") BytesNode aBytes,
+                @Cached("create()") BytesNode bBytes) {
+            return a.hashesMatch(b) && a.byteLength() == b.byteLength() && Arrays.equals(aBytes.execute(a), bBytes.execute(b));
```

Comment:
This is not going to perform so well.
I think we should compare byteLengths first, then only if available compare the Hash and otherwise have our own compare loop, all of this profiled, as in the StringEqualNode.